### ISSUE TITLE
refresh() is confusing

### ISSUE THREAD

**Issue by mottosso** [pyblish/pyblish-qml#167]
I just had a look at this and can't believe I let this method slip through. :O

- [control.py:refresh()](https://github.com/pyblish/pyblish-qml/blob/95444f93a69704cee764175769620d772355899a/pyblish_qml/control.py#L528)

There are a number of issues with this method;

1. It is inefficient, running in the main thread where it really should be running asynchronously like it's siblings.
1. It is inefficient. Calling `host.context()` twice, each time is a *significant* performance bottleneck, and nothing will have changed after the first call.
1. It is needless. `host.context()` was already called in the calling function, and nothing will have changed since.
1. It is confusing; re-ordering the model what seems like multiple times.
1. It is only used once; it should probably not even be a method on the class.

I can't begin to describe how difficult this logic will be to maintain (I already don't understand it), and I'm astonished it hasn't already broken things or made itself more visible than it has.

Looking at the [source commit](https://github.com/pyblish/pyblish-qml/blame/95444f93a69704cee764175769620d772355899a/pyblish_qml/control.py#L528), I can see @tokejepsen's name; just be assured that you've done nothing wrong, we just need to take more care in our pull requests and make sure we make tests and properly profile the changes before merging.

Just so we're not losing functionality, could you remind me what this method does?

**Comment by tokejepsen** [comment]
This was related to [context reordering](http://forums.pyblish.com/t/gui-context-reordering/213). It refreshes the context from the host, because the GUI was adding instances as they get added to the context and if a collector changed that order, it wasn't reflected in the GUI.
This only happens on collection, and should happen on processing the plugins.

**Comment by tokejepsen** [comment]
> This only happens on collection, and should happen on processing the plugins.

Correction: This only happens on collection, and shouldn't happen on processing the plugins.

**Comment by mottosso** [comment]
Ok, I'll keep this behaviour in the forthcoming fix.

**Comment by BigRoy** [comment]
Even though this is altered in #168 it's not really simplified? Or did the problem just shift? `refresh()` is still rather complicated and doing multiple calls to `host.context()`? It only became async?

Looking over the code of how it actually reorders it kind of felt beyond me. E.g. what is the plugin_index and why is that offset there? Having no clue what the code does and making assumptions about it... I came up with this:

```python
# pseudocode
    def refresh(self):

        self.item_model.beginResetModel()

        # preserve the plugin index offset
        # we don't want to reorder before that
        plugin_index = len(self.item_model.plugins)

        items = self.item_model.items
        plugin_items = items[plugin_index:]
        reorder = dict()    #  (new index, item)

        # compute reorder mapping taking the index
        # from the host order and map it to the item
        # in the item model
        context = self.host.contex()
        for i, instance in enumerate(context):
            index = plugin_index + i
            for item in plugin_items:
                if item.id == str(instance):
                    reorder[index] = item
                    break

        # reorder the plug-ins in item model
        for index, item in reoder.items():
            self.item_model.items[index] = item

        self.item_model.endResetModel()
```

Basically it's somewhat of a rewrite supposed to be more optimized of what I'm assuming the method is trying to achieve. Doing less calls to properties and methods changing it to mostly local aritmethics. Note the pseudocode header comment.

**Comment by mottosso** [comment]
Well spotted, I had actually forgotten to remove the method. Have a look now.

**Comment by mottosso** [comment]
Here is the reordering part.

- https://github.com/pyblish/pyblish-qml/pull/168/files#diff-ac156cba73ac1ef385b357bbfc21d258R615

**Comment by BigRoy** [comment]
Thanks! Much clearer now. Seems like the `plugin_index` before was to preserve the "Context" object. 

By the way... what happens if an instance is named Context? Does this bug out?

**Comment by mottosso** [comment]
> By the way... what happens if an instance is named Context? Does this bug out?

Haha, I don't know. Possibly.

> Seems like the plugin_index before was to preserve the "Context" object.

I think it was because instances and plug-ins were stored together in the same model, and he was finding at which index plug-ins started, and worked from there. The problem with that is that they aren't guaranteed to lie next to each other, but can lie one after the other in any order.
